File: QualityTrim.R
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Title :  QualityTrim.R 
# Version : 1.0
#
# Purpose : A tool that trims the 3' end of reads using sliding window average Q-score threshold
#  
# Version Notes : 
#
# Created.date  : 27 Apr 2016
# Created.by    : Dan Spakowicz
# Updated.date  : 05 May 2016 
# Updated.by    : DS
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Usage: Rscript QualityTrim.py -i <input fastq file> -s <qscore file> -t <threshold cutoff score> -o <output file type {fastq, fasta, both}> -m <minimum read length>
### Example: Rscript QualityTrim.py -i input.txt -s qscore.txt -t 25 -o fastq -m 10 > outputfile.txt

### Notes: Requires input file in fastq format and score file in "|" separated format.
###        If no threshold cutoff score is specified the default of 40 is used.
###        If no output file type is specified the default fastq is used.
###        If no minimum read length is specified, no minimum is used.
###        To save output files, use > outputfile.txt 


rm(list=ls())
oldw <- getOption("warn")
options(warn = -1)

# Load the required packages
list.of.packages <- c("optparse")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library(optparse)

# set arguments
option_list = list(
  make_option(c("-i", "--input"), type="character", default=NULL, 
              help="input fastq file name", metavar="character"),
  make_option(c("-s", "--score"), type="character", default="./qscores.txt", 
              help="score file name", metavar="character"),
  make_option(c("-t", "--threshold"), type="character", default=20, 
              help="threshold cutoff score", metavar="character"),
  make_option(c("-m", "--minimum"), type="character", default=0, 
              help="minimum read length", metavar="character"),
  make_option(c("-o", "--out"), type="character", 
              default="fastq", 
              help="output file type, must be one of the following: {fastq, fasta, both}, [default= %default]", 
              metavar="character")
); 

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

# infile <- "example_100.fastq"
# scorefile <- "qscores.txt"
# threshold <- 20
# minimum <- 20
# outputtype <- "fastq"

# Trim function
Trim <- function(infile, scorefile, threshold, minimum, outputtype) {
  
  # read in and format fastq file
  input <- readLines(infile)
  fastq.split <- strsplit(as.character(input), split = "")
  
  # read in the quality scores file
  scorefile <- read.table(scorefile, sep = "|", quote = "", comment.char = "")
  
  # list of just the qual score lines
  qualLines <- seq(4, length(input), by = 4)
  qualscores <- fastq.split[qualLines]
  
  # convert ascii qual scores to phred scores
  phred.scores <- vector("list", length(qualscores))
  for ( i in 1:length(qualscores)) {
    for (j in 1:length(qualscores[[i]])){
      phred.scores[[i]][j] <- scorefile$V2[which(qualscores[[i]][j]==scorefile$V1)]
    }
  }
  
  # sliding window average over the phred scores
  cut.logical <- vector("list", length(phred.scores))
  for ( i in 1:length(phred.scores)) {
    for (j in 1:length(phred.scores[[i]])){
      cut.logical[[i]][j] <- sum(phred.scores[[i]][j:(j+3)])/4 < threshold
        }
    }
  
  # find the location of the first average less than the threshold (TRUE)
  cut.loc <- vector("list", length(cut.logical))
  for ( i in 1:length(cut.logical)) {
    cut.loc[[i]] <- min(which(cut.logical[[i]] == TRUE))
    if(cut.loc[[i]] == Inf){cut.loc[[i]] = length(cut.logical[[i]])}
  }
 
  # trim sequence and qual scores
  cut.loc.vec <- unlist(cut.loc)
  for (i in 1:length(cut.loc.vec)){
    fastq.split[[(i*4-2)]] <- fastq.split[[(i*4-2)]][1:(cut.loc.vec[i]+1)]
    fastq.split[[(i*4)]] <- fastq.split[[(i*4)]][1:(cut.loc.vec[i]+1)]
  }
  
  # remove sequences shorter than the minimum length
  minlength <- vector("list", length(qualLines))
  for (i in qualLines){
    minlength[[i]] <- length(fastq.split[[i]]) > minimum
  }
  for (i in qualLines){
    minlength[[(i-3)]] <- minlength[[i]]
    minlength[[(i-2)]] <- minlength[[i]]
    minlength[[(i-1)]] <- minlength[[i]]
  }
  minlength.vec <- unlist(minlength)
  fastq.split.min <- fastq.split[minlength.vec]
  
  # output as fastq, fasta or both
  if ( outputtype == "fasta" ){
    
    a <- seq(from = 1, to = length(fastq.split.min), by = 4)
    b <- seq(from = 2, to = length(fastq.split.min), by = 4)               
    fastaLines <- c( matrix(c(a,b), nrow=2, byrow=TRUE) )
    fasta <- fastq.split.min[fastaLines]
    for ( i in 1: length(fasta)){
      if(fasta[[i]][1]=="@"){
        fasta[[i]][1] <- ">"
      }
    }
    fileFasta<-file("output.fasta")
    writeLines(unlist(lapply(fasta, paste, collapse="")), fileFasta)
    close(fileFasta)
  }
  
  if ( outputtype == "fastq" ){
    fileFastq<-file("output_R.fastq")
    writeLines(unlist(lapply(fastq.split.min, paste, collapse="")), fileFastq)
    close(fileFastq)
  }
  if ( outputtype == "both" ){
    fileFastq<-file("output_R.fastq")
    writeLines(unlist(lapply(fastq.split.min, paste, collapse="")), fileFastq)
    close(fileFastq)
    a <- seq(from = 1, to = length(fastq.split.min), by = 4)
    b <- seq(from = 2, to = length(fastq.split.min), by = 4)               
    fastaLines <- c( matrix(c(a,b), nrow=2, byrow=TRUE) )
    fasta <- fastq.split.min[fastaLines]
    for ( i in 1: length(fasta)){
      if(fasta[[i]][1]=="@"){
        fasta[[i]][1] <- ">"
      }
    }
    fileFasta<-file("output_R.fasta")
    writeLines(unlist(lapply(fasta, paste, collapse="")), fileFasta)
    close(fileFasta)
  }
}

Trim(opt$input, opt$score, opt$threshold, opt$minimum, opt$out)

options(warn = oldw)  

File: QualityTrim.py
### Title: QualityTrim.py
### Version: 1.0
### Purpose: A tool that trims the 3' end of reads using sliding window average Q-score threshold
### Created Date: 01 May 2016
### Created By: Heather Wells

### Usage: python QualityTrim.py -i <input fastq file> -s <qscore file> -t <threshold cutoff score> -m <minimum read length> -f <output file type {fastq, fasta, both}> -o <output file name>
### Example: python QualityTrim.py -i input.txt -s qscore.txt -t 25 -o fastq -m 10 > outputfile.txt

### Notes: Requires input file in fastq format and score file in "|" separated format.
###        If no threshold cutoff score is specified the default of 40 is used.
###        If no output file type is specified the default fastq is used.
###        If no minimum read length is specified, no minimum is used.
###        To save output files, use >outputfile.txt 

#import arguments
import argparse
parser=argparse.ArgumentParser(description='QualityTrim')
parser.add_argument('-i', '--input', help='input file name', required=True)
parser.add_argument('-s', '--score', help='score file name', required=True)
parser.add_argument('-t', '--threshold', help='threshold cutoff score', default=20, type=int)
parser.add_argument('-w', '--window', help='window size', default=4, type=int)
parser.add_argument('-m', '--minimum', help='minimum read length', default=0, type=int)
parser.add_argument('-f', '--outputtype', help='output file type, must be one of the following: {fastq, fasta, both}', default='fastq')
parser.add_argument('-o', '--outputname', help='output file name', default="output")
args=parser.parse_args()

import re
import numpy


def Trim(inputfile,scorefile,threshold,outputtype):

    #open and save inputs
    with open(args.input) as input:
        input=input.read().splitlines()[0:100]
    with open(args.score) as scores:
        scores=numpy.loadtxt(scores,delimiter="|",dtype=str,comments=None)

    #initialize empty output lists
    fastq=[]
    fasta=[]

    #for loop with step size of 4 to analyze one sequence chunk at a time
    for i in range(0,len(input),4):
        
        #the ascii qscore is the 4th (i+3) of each chunk of 4 lines
        q_ascii_i=list(input[i+3])

        #convert ascii code into numeric scores
        #for each character, find the match in the scores array and add to q_score_i list
        #then, for each q_score_i list, average the scores 4 at a time (sliding window)
        #if the average is less than the specified or default (20) threshold, break the loop and store the cut value as the window midpoint
        #if the average is greater than the default for the entire loop, set cut to length of q_score_i (keeps entire sequence)
        #the sequence is the 2nd (i+1) of each chunk of 4 lines, only keep sequence up to the cut point
        q_score_i=[]
        window=args.window
        for each in q_ascii_i:
            q_score_i.extend(map(int,scores[numpy.where(scores==each)[0],1]))
        for j in range(len(q_score_i)-window):
            avg=sum(q_score_i[j:j+window])/window
            if avg<args.threshold:
                cut=j+window/2
                break
            else: cut=len(q_score_i)
        seq_i=input[i+1][0:cut]
        
        #extend fasta and fastq file types with correct format only if length minimum is met (defaults with no min)
        if len(seq_i)>args.minimum:
            fastq.extend(input[i]+"\n"+seq_i+"\n+\n"+input[i+3]+"\n")
            fasta.extend(">"+input[i]+"\n"+seq_i+"\n")

    fastq="".join(fastq)
    fasta="".join(fasta)

    #save user-specified or default (fastq) output
    if args.outputtype=='fastq':
        fastq_file=open(args.outputname+".fastq","w")
        fastq_file.write(fastq)
        fastq_file.close()
    if args.outputtype=='fasta':
        fasta_file=open(args.outputname+".fasta","w")
        fasta_file.write(fasta)
        fasta_file.close()
    if args.outputtype=='both':
        fastq_file=open(args.outputname+".fastq","w")
        fastq_file.write(fastq)
        fastq_file.close()
        fasta_file=open(args.outputname+".fasta","w")
        fasta_file.write(fasta)
        fasta_file.close()

Trim(args.input,args.score,args.threshold,args.outputtype)
